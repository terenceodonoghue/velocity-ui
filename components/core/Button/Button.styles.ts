import { css, FunctionInterpolation, Theme } from '@emotion/react';
import { darken } from 'polished';

export const button: FunctionInterpolation<Theme> = ({
  transitions,
  typography,
}) => css`
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  border-radius: 4px;
  height: 40px;
  padding: 0;
  width: 200px;
  font-weight: ${typography.fontWeight.medium};
  cursor: pointer;
  transition: background-color ${transitions.duration.standard}ms
    ${transitions.easing.easeInOut};

  &:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  &:focus {
    outline: 0;
  }
`;

export const contained: FunctionInterpolation<Theme> = ({ palette }) => css`
  background-color: ${palette.accent};
  color: ${palette.neutral[50]};

  &:enabled:hover {
    background-color: ${darken(palette.tonalOffset, palette.accent)};
  }
`;

export const outlined: FunctionInterpolation<Theme> = ({ palette }) => css`
  background: ${palette.primary};
  color: ${palette.accent};

  &:enabled:hover {
    background-color: ${darken(palette.tonalOffset, palette.accent)};
  }
`;

export const text: FunctionInterpolation<Theme> = ({ palette }) => css`
  background: none;
  color: ${palette.accent};

  &:enabled:hover {
    background-color: ${darken(palette.tonalOffset, palette.accent)};
  }
`;
