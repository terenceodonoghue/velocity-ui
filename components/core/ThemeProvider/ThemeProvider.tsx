import React, { FunctionComponent } from 'react';
import { ThemeProvider as EmotionProvider, Theme } from '@emotion/react';
import merge from 'lodash/merge';
import { rgba } from 'polished';
import type { ThemeProps } from './ThemeProvider.types';

export const defaultTheme: Theme = {
  palette: {
    primary: rgba('#2e5bff', 0.2),
    secondary: '#e0e7ff',
    accent: '#2e5bff',
    neutral: {
      50: '#fff',
      100: '#f4f6fc',
      200: '#eef3f5',
      300: '#e4e8f0',
      400: '#bfc5d2',
      500: '#b0bac9',
      600: '#8798ad',
      700: '#8097b1',
      800: '#69707f',
      900: '#2E384D',
    },
    ui: {
      green: '#33ac2e',
      purple: '#8c54ff',
      red: '#d63649',
      teal: '#00c1d4',
      yellow: '#f7c137',
    },
    tonalOffset: 0.2,
  },
  transitions: {
    duration: {
      shortest: 150,
      shorter: 200,
      short: 250,
      standard: 300,
      complex: 375,
      enteringScreen: 225,
      leavingScreen: 195,
    },
    easing: {
      easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
      easeOut: 'cubic-bezier(0.0, 0, 0.2, 1)',
      easeIn: 'cubic-bezier(0.4, 0, 1, 1)',
      sharp: 'cubic-bezier(0.4, 0, 0.6, 1)',
    },
  },
  typography: {
    fontFamily: ['Rubik', 'sans-serif'].join(','),
    fontSize: 15,
    fontWeight: {
      light: 300,
      regular: 400,
      medium: 500,
      bold: 700,
    },
  },
  zIndex: {
    appBar: 1100,
    drawer: 1200,
  },
};

const ThemeProvider: FunctionComponent<ThemeProps> = ({ children, theme }) => (
  <EmotionProvider theme={theme ? merge(defaultTheme, theme) : defaultTheme}>
    {children}
  </EmotionProvider>
);

export default ThemeProvider;
