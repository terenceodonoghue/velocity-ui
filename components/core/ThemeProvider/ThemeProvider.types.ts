export interface Theme {
  palette?: {
    primary?: string;
    secondary?: string;
    accent?: string;
    neutral?: {
      50?: string;
      100?: string;
      200?: string;
      300?: string;
      400?: string;
      500?: string;
      600?: string;
      700?: string;
      800?: string;
      900?: string;
    };
    ui?: {
      green?: string;
      purple?: string;
      red?: string;
      teal?: string;
      yellow?: string;
    };
    tonalOffset?: number;
  };
  transitions?: {
    duration?: {
      shortest?: number;
      shorter?: number;
      short?: number;
      standard?: number;
      complex?: number;
      enteringScreen?: number;
      leavingScreen?: number;
    };
    easing?: {
      easeInOut?: string;
      easeOut?: string;
      easeIn?: string;
      sharp?: string;
    };
  };
  typography?: {
    fontFamily?: string;
    fontSize?: number;
    fontWeight?: {
      light?: number;
      regular?: number;
      medium?: number;
      bold?: number;
    };
  };
  zIndex?: {
    appBar?: number;
    drawer?: number;
  };
}

export interface ThemeProps {
  theme?: Theme;
}
