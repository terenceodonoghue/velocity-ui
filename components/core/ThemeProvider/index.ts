export { default, defaultTheme } from './ThemeProvider';
export type { Theme } from './ThemeProvider.types';
