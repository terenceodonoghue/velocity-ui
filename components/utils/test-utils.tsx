import React, { FunctionComponent, ReactElement } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import ThemeProvider from '~/components/core/ThemeProvider';

const Providers: FunctionComponent = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

export { customRender as render };
