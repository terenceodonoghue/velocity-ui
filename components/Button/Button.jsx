/** @jsx jsx */
import { jsx } from '@emotion/core';
import * as S from './Button.styles';

const Contained = props => <S.Contained {...props} />;
const Outlined = props => <S.Outlined {...props} />;
const Text = props => <S.Text {...props} />;
const Fab = props => <S.Fab {...props} />;

export default {
  Contained,
  Outlined,
  Text,
  Fab,
};
