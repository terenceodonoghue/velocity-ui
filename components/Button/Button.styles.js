import { css } from '@emotion/core';
import styled from '@emotion/styled';

const base = css`
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  padding: 0;
  font-size: var(--base-font-size);
  font-weight: var(--font-weight-medium);
  cursor: pointer;
  transition: var(--transition);

  &:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  &:focus {
    outline: 0;
  }
`;

export const Contained = styled.button`
  ${base};
  border-radius: 4px;
  height: 40px;
  width: ${props => props.width || '200px'};
  background-color: var(--color-clear-blue);
  color: white;

  &:enabled:hover {
    background-color: var(--color-clear-blue-dark);
  }
`;

export const Outlined = styled.button`
  ${base};
  border-radius: 4px;
  height: 40px;
  width: ${props => props.width || '200px'};
  background: rgba(46, 91, 255, 0.2);
  color: var(--color-clear-blue);

  &:enabled:hover {
    background-color: rgba(46, 91, 255, 0.3);
  }
`;

export const Text = styled.button`
  ${base};
  border-radius: 4px;
  height: 40px;
  width: ${props => props.width || '200px'};
  background: none;
  color: var(--color-clear-blue);

  &:enabled:hover {
    background-color: rgba(46, 91, 255, 0.1);
  }
`;

export const Fab = styled.button`
  ${base};
  border-radius: 50%;
  height: 70px;
  width: 70px;
  background-color: var(--color-clear-blue);
  color: white;

  &:enabled:hover {
    background-color: var(--color-clear-blue-dark);
  }
`;
