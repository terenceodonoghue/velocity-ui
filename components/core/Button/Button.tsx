import React, { forwardRef } from 'react';
import * as css from './Button.styles';
import type { ButtonProps } from './Button.types';

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ variant = 'contained', ...props }, ref) => (
    <button
      css={[css.button, css[variant]]}
      data-testid="button"
      ref={ref}
      type="button"
      {...props}
    />
  ),
);

export default Button;
