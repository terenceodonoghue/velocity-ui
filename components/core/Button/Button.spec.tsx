import React from 'react';
import { render } from 'test-utils';
import Button from './Button';
import { defaultTheme } from '../ThemeProvider';

describe('Button', () => {
  it('is contained by default', () => {
    const { container, getByTestId } = render(<Button />);

    expect(getByTestId('button')).toHaveStyle({
      backgroundColor: defaultTheme.palette.accent,
      color: defaultTheme.palette.neutral[50],
    });
    expect(container).toMatchSnapshot();
  });

  it('renders contained variant', () => {
    const { container, getByTestId } = render(<Button variant="contained" />);

    expect(getByTestId('button')).toHaveStyle({
      backgroundColor: defaultTheme.palette.accent,
      color: defaultTheme.palette.neutral[50],
    });
    expect(container).toMatchSnapshot();
  });

  it('renders outlined variant', () => {
    const { container, getByTestId } = render(<Button variant="outlined" />);

    expect(getByTestId('button')).toHaveStyle({
      backgroundColor: defaultTheme.palette.primary,
      color: defaultTheme.palette.accent,
    });
    expect(container).toMatchSnapshot();
  });

  it('renders text variant', () => {
    const { container, getByTestId } = render(<Button variant="text" />);

    expect(getByTestId('button')).toHaveStyle({
      background: 'none',
      color: defaultTheme.palette.accent,
    });
    expect(container).toMatchSnapshot();
  });
});
