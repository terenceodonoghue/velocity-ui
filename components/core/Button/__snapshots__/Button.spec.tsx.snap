// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button is contained by default 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: none;
  border-radius: 4px;
  height: 40px;
  padding: 0;
  width: 200px;
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
  background-color: #2e5bff;
  color: #fff;
}

.emotion-0:disabled {
  opacity: 0.3;
  pointer-events: none;
}

.emotion-0:focus {
  outline: 0;
}

.emotion-0:enabled:hover {
  background-color: #002bc7;
}

<div>
  <button
    class="emotion-0"
    data-testid="button"
    type="button"
  />
</div>
`;

exports[`Button renders contained variant 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: none;
  border-radius: 4px;
  height: 40px;
  padding: 0;
  width: 200px;
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
  background-color: #2e5bff;
  color: #fff;
}

.emotion-0:disabled {
  opacity: 0.3;
  pointer-events: none;
}

.emotion-0:focus {
  outline: 0;
}

.emotion-0:enabled:hover {
  background-color: #002bc7;
}

<div>
  <button
    class="emotion-0"
    data-testid="button"
    type="button"
  />
</div>
`;

exports[`Button renders outlined variant 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: none;
  border-radius: 4px;
  height: 40px;
  padding: 0;
  width: 200px;
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
  background: rgba(46,91,255,0.2);
  color: #2e5bff;
}

.emotion-0:disabled {
  opacity: 0.3;
  pointer-events: none;
}

.emotion-0:focus {
  outline: 0;
}

.emotion-0:enabled:hover {
  background-color: #002bc7;
}

<div>
  <button
    class="emotion-0"
    data-testid="button"
    type="button"
  />
</div>
`;

exports[`Button renders text variant 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: none;
  border-radius: 4px;
  height: 40px;
  padding: 0;
  width: 200px;
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
  background: none;
  color: #2e5bff;
}

.emotion-0:disabled {
  opacity: 0.3;
  pointer-events: none;
}

.emotion-0:focus {
  outline: 0;
}

.emotion-0:enabled:hover {
  background-color: #002bc7;
}

<div>
  <button
    class="emotion-0"
    data-testid="button"
    type="button"
  />
</div>
`;
